### PR TITLE
Fix AOTI cond predicate unbacked symbol dependency

### LINKED ISSUE #140842: torch.cond with unbacked symint failed to compile in AOTI
### 🐛 Describe the bug

Repro:

```python
def test_cond_symint_input(self):
        class M(torch.nn.Module):
            def forward(self, x, y, z):
                a = y.shape[0]
                b = z.item()
                def true_fn(x):
                    return x + a 

                def false_fn(x):
                    return x + b * z

                return torch.cond(x.shape[0] > 5, true_fn, false_fn, (x,))
        
        inp = (torch.ones(3, 3), torch.ones(3, 3), torch.tensor(5))
        ep = torch.export.export(M(), inp, dynamic_shapes={"x": {0: Dim("d")}, "y": {0: Dim("d1")}, "z": None})
        print(ep)
        path = torch._inductor.aot_compile(ep.module(), inp)
        breakpoint()
        print(f"{path[:-3]}.cpp")
```

Error message: https://gist.github.com/larryliu0820/b2c4f5da466462ad53c6fcfbe9a4c4bc
Generated cpp: https://gist.github.com/larryliu0820/5423934533003b891425ad75cc32a0b0

### Versions

```
Collecting environment information...
PyTorch version: 2.6.0.dev20241115+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-2)
Clang version: Could not collect
CMake version: version 3.30.0
Libc version: glibc-2.34

Python version: 3.11.9 (main, Apr 19 2024, 16:48:06) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.4.3-0_fbk14_hardened_2601_gcd42476b84e9-x86_64-with-glibc2.34
Is CUDA available: False
CUDA runtime version: 12.0.140
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA H100
Nvidia driver version: 550.90.07
cuDNN version: Probably one of the following:
/usr/lib64/libcudnn.so.9.1.0
/usr/lib64/libcudnn_adv.so.9.1.0
/usr/lib64/libcudnn_cnn.so.9.1.0
/usr/lib64/libcudnn_engines_precompiled.so.9.1.0
/usr/lib64/libcudnn_engines_runtime_compiled.so.9.1.0
/usr/lib64/libcudnn_graph.so.9.1.0
/usr/lib64/libcudnn_heuristic.so.9.1.0
/usr/lib64/libcudnn_ops.so.9.1.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      52 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             46
On-line CPU(s) list:                0-45
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 9654 96-Core Processor
CPU family:                         25
Model:                              17
Thread(s) per core:                 1
Core(s) per socket:                 46
Socket(s):                          1
Stepping:                           1
BogoMIPS:                           4792.78
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean pausefilter pfthreshold v_vmsave_vmload vgif vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid fsrm flush_l1d arch_capabilities
Virtualization:                     AMD-V
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          2.9 MiB (46 instances)
L1i cache:                          2.9 MiB (46 instances)
L2 cache:                           23 MiB (46 instances)
L3 cache:                           736 MiB (46 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-45
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:           Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] executorch==0.5.0a0+b7763e9
[pip3] flake8==6.1.0
[pip3] flake8-breakpoint==1.1.0
[pip3] flake8-bugbear==23.9.16
[pip3] flake8-comprehensions==3.14.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-plugin-utils==1.3.3
[pip3] flake8-pyi==23.5.0
[pip3] flake8-simplify==0.19.3
[pip3] mypy==1.10.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.2
[pip3] nvidia-cublas-cu12==12.1.3.1
[pip3] nvidia-cuda-cupti-cu12==12.1.105
[pip3] nvidia-cuda-nvrtc-cu12==12.1.105
[pip3] nvidia-cuda-runtime-cu12==12.1.105
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.0.2.54
[pip3] nvidia-curand-cu12==10.3.2.106
[pip3] nvidia-cusolver-cu12==11.4.5.107
[pip3] nvidia-cusparse-cu12==12.1.0.106
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.6.20
[pip3] nvidia-nvtx-cu12==12.1.105
[pip3] optree==0.12.1
[pip3] pytorch-triton==3.1.0+cf34004b8a
[pip3] torch==2.6.0.dev20241115+cpu
[pip3] torchao==0.5.0+git0916b5b2
[pip3] torchaudio==2.5.0.dev20241112+cpu
[pip3] torchsr==1.0.4
[pip3] torchtune==0.4.0.dev20241111+cpu
[pip3] torchvision==0.20.0.dev20241112+cpu
[conda] executorch                0.5.0a0+b7763e9          pypi_0    pypi
[conda] numpy                     1.23.2                   pypi_0    pypi
[conda] nvidia-cublas-cu12        12.1.3.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.1.105                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.0.2.54                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.2.106               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.4.5.107               pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.1.0.106               pypi_0    pypi
[conda] nvidia-nccl-cu12          2.21.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.6.20                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.1.105                 pypi_0    pypi
[conda] optree                    0.12.1                   pypi_0    pypi
[conda] pytorch-triton            3.1.0+cf34004b8a          pypi_0    pypi
[conda] torch                     2.6.0.dev20241115+cpu          pypi_0    pypi
[conda] torchao                   0.5.0+git0916b5b2          pypi_0    pypi
[conda] torchaudio                2.5.0.dev20241112+cpu          pypi_0    pypi
[conda] torchfix                  0.5.0                    pypi_0    pypi
[conda] torchsr                   1.0.4                    pypi_0    pypi
[conda] torchtune                 0.4.0.dev20241111+cpu          pypi_0    pypi
[conda] torchvision               0.20.0.dev20241112+cpu          pypi_0    pypi
```

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #186094

AOTI lowering for torch.cond can store its predicate as a SymPy boolean
relation inside ExternKernel.constant_args, for example `u0 > 0`. The scheduler
uses ExternKernel.get_free_symbol_uses() to add dependencies from users of
unbacked symbols to their defining nodes, but the constant-argument scanner
only recognized SymTypes and sympy.Expr.

SymPy boolean relations are sympy.Basic but not sympy.Expr, so the cond node did
not report its use of `u0`. Scheduler DCE could remove the dynamic scalar node
that defined `u0`, while C++ wrapper codegen still emitted a predicate such as
`if (u0 > 0L)`, producing generated C++ with an undeclared symbol. Issue
#139798, which isolates this predicate case, was closed as a duplicate of
#140842.

Accept sympy.Basic in maybe_free_symbols() and
maybe_free_unbacked_symbols(). This keeps the normal scheduler dependency path
responsible for preserving only definitions that are actually used. The
earlier alternative of pinning every node that defines an unbacked symbol
would mask missing dependencies and retain unrelated dead nodes.

The regression runs both predicate branches through the current AOTI package
runner. Automatically copied arrayref stack-allocation and lazy dual-wrapper
variants remain unsupported for independent wrapper limitations, so register
them as expected failures rather than skips.

Fixes #140842

Generated by my agent

Benchmark Results:

A small Python microbenchmark compared the previous scanner predicate
(SymTypes, Expr) with the new predicate (SymTypes, sympy.Basic) over 200000
calls:

- expr: old 0.432370s, new 0.428872s, same symbols found
- relational predicate: old 0.054814s and missed `u0`, new 0.295944s and found `u0`
- nested tuple/list: old 0.921164s, new 1.300788s, same expression symbols plus relational support

The additional work is limited to compile-time scanning of SymPy Basic objects
that the old code skipped. Generated runtime code is unchanged except that the
required symbol definition is retained.

Test Plan:

- `python test/inductor/test_aot_inductor.py AOTInductorTestABICompatibleCpu.test_cond_unbacked_symint_predicate_cpu AOTInductorTestABICompatibleGpu.test_cond_unbacked_symint_predicate_cuda AOTInductorTestDualWrapper.test_cond_unbacked_symint_predicate_cuda` (2 passed, 1 expected failure)
- `python test/inductor/test_aot_inductor_arrayref.py AOTInductorTestABICompatibleCpuWithStackAllocation.test_cond_unbacked_symint_predicate_cpu_with_stack_allocation` (1 expected failure)
- `python test/inductor/test_aot_inductor.py AOTInductorTestABICompatibleCpu.test_cond_unbacked_symint_closure_dynamic_False_cpu AOTInductorTestABICompatibleCpu.test_cond_unbacked_symint_closure_dynamic_True_cpu AOTInductorTestABICompatibleCpu.test_cond_symint_input_cpu` (3 passed)
- `git diff --check`
- `lintrunner -a`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says Dr.CI's new failures were all from the copied arrayref stack-allocation variant of the new AOTI cond regression. I reproduced the C++ compile error locally, then skipped only that unsupported copied stack-allocation variant in test_aot_inductor_arrayref.py while keeping the main regression test active.

**Comment by jansel** [comment]
My agent says the remaining torchtitan_features_integration failure is the known torchcomms FSDP+TP+PP+compile issue tracked in #186031, with the same `Could not resolve the process group registered under the name` traceback. This is unrelated to this AOTI/Inductor symbol dependency fix, so I removed the extra `ciflow/torchtitan` label while leaving the relevant Inductor CI coverage in place.

**Review by larryliu0820 — APPROVED** [review]
Nice

**Review by desertfire — APPROVED** [review]

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main to replace the stale Dr.CI run. Current main now copies AOTI tests into lazy dual-wrapper coverage, so I moved the regression to the current package runner and registered the independently unsupported dual-wrapper and arrayref copies as expected failures rather than skips. The focused CPU and GPU regressions pass, the copied variants report expected failures, and lintrunner passes.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says ghstack restored the obsolete ciflow/torchtitan label during submission, so I removed it again; the known torchcomms failure remains outside this PR. I also corrected the retained PR description to match the amended commit and removed the unrelated Fixes #137520 claim.

**Comment by jansel** [comment]
My agent says pytorch-bot immediately re-added ciflow/torchtitan after my removal, so this label is currently policy-enforced for the changed files. The refreshed run will include TorchTitan coverage; any recurrence of the exact torchcomms failure remains tracked by #186031.

**Review by mlazos — APPROVED** [review]

**Comment by jansel** [comment]
My agent says the two new CI failures were unexpected successes from the stale lazy dual-wrapper xfail; I removed that xfail and the test now passes unmasked. I also addressed Claude’s follow-ups by adding a normal Python-wrapper regression (which reproduces the missing-symbol NameError when the fix is reverted), sharing it with the AOTI test, and correcting the symbol-scanner comment. The branch is rebased onto current main; focused CPU/GPU/AOTI tests and lintrunner pass.
